### PR TITLE
docs: add SQL plan management with parameter partition

### LINKED ISSUE #15074: draft a proposal about how to support multiple plan for one fingerprint in SPM
## Feature Request

**Is your feature request related to a problem? Please describe:**

For now, all the baseline in SPM are optimized for only one parameter of a normalized SQL. It can led to a sub-optimal plan when another parameter is presented by the user.

**Describe the feature you'd like:**

Find a workaround to solve this issue. Seems we can consider oracle adaptive cursor sharing.

**Describe alternatives you've considered:**

No

**Teachability, Documentation, Adoption, Migration Strategy:**

For oracle adaptive cursor sharing, these documents may help:
- [Update on Adaptive Cursor Sharing](https://blogs.oracle.com/optimizer/update-on-adaptive-cursor-sharing)
- [Adaptive Cursors and SQL Plan Management](https://www.oracle.com/technical-resources/articles/database/sql-11g-sqlplanmanagement.html)

### PR DESCRIPTION
**Pull request by lzmhhh123**
### What problem does this PR solve? 
fix #15074 

### What is changed and how it works?
Add a draft of SQL plan management with parameter partition.

### CODE REVIEW

File: docs/design/2020-03-03-SPM-with-parameter-partition.md [hunk 1]
```diff
@@ -0,0 +1,109 @@
...
+binding type   |   Priority    | way to create             | action scope
+---------------|:------------- |:--------------------------|-------------------------------------------------
+SQL bind       | first choice  | use SQL                   | one standardized SQL
+Parameter bind | second choice | enable SPM plan evolution | one standardized SQL with some parameters' range
+
+The parameters' range is the range of the value of the constant which in "single column rangeable" condition.
+
+## Compatibility
+
+Incompatible with older versions of **SQL Bind**.
```
Comment (zz-jason) [inline docs/design/2020-03-03-SPM-with-parameter-partition.md]: We should find a way to make it compatible with SQL Bind in release 3.0. If there is not such method, we should provide an upgrade guide on how to migrate from old SQL Bind to the new SQL Bind.

Comment (lzmhhh123) [reply docs/design/2020-03-03-SPM-with-parameter-partition.md]: Oh, it's the fault of the document. Actually, it's compatible with SQL Bind in release 3.0.

File: docs/design/2020-03-03-SPM-with-parameter-partition.md [hunk 2]
```diff
@@ -0,0 +1,109 @@
...
+
+TiDB optimizer uses [Conjunctive Normal Form](https://en.wikipedia.org/wiki/Conjunctive_normal_form) (hereinafter CNF) to divide `where` conditions. For example:
+
+```sql
+where t.a = 1 and 
+      t.b >= 2 and 
+      (t.c*2 == 1 or t.c is null)
+```
+
+Its CNF is $(`t.a=1`) \cap (`t.b>=2`) \cap (`t.c*2==1` \cup `t.c is null`)$. For each single condition, "single column rangeable" is like the format as **${column_name} ${compare_operator} constant**. In this example, the first and second conditions are "single column rangeable".
```
Comment (eurekaka) [inline docs/design/2020-03-03-SPM-with-parameter-partition.md]: What does `compare_operator` mean?

File: docs/design/2020-03-03-SPM-with-parameter-partition.md [hunk 3]
```diff
@@ -0,0 +1,109 @@
...
+		INDEX sql_index(original_sql(1024), default_db(1024)) COMMENT "accelerate the speed when add global binding query",
+		INDEX time_index(update_time) COMMENT "accelerate the speed when querying with last update time"
+	) ENGINE=InnoDB DEFAULT CHARSET=utf8mb4 COLLATE=utf8mb4_bin;
+```
+
+- Add another parameter binding cache in TiDB
+
+- Change the binding chosen logic
+    1. Reserve the origin SQL bind chosen logic.
+    2. If hits none SQL bind, the try to hit parameter bind.
```
Comment (eurekaka) [inline docs/design/2020-03-03-SPM-with-parameter-partition.md]: How could it happen that no SQL bind is hit?

File: docs/design/2020-03-03-SPM-with-parameter-partition.md [hunk 4]
```diff
@@ -0,0 +1,109 @@
...
+
+- SPM plan evolution
+    1. Only change the parameter bind, not affect SQL bind.
+    2. Insert into the bucket like insert histogram bucket in statistics.
```
Comment (eurekaka) [inline docs/design/2020-03-03-SPM-with-parameter-partition.md]: Each bucket would have more than one binding? how is the evolution triggered? how to verify the bindings in each bucket?

File: docs/design/2020-03-03-SPM-with-parameter-partition.md [hunk 5]
```diff
@@ -0,0 +1,109 @@
...
+    3. Try to split the bucket if its count exceeds twice count of the smallest bucket.
+    4. If the number of bind SQL for one bucket is more than 20 or another number, drop the binding with max execute_time.
```
Comment (eurekaka) [inline docs/design/2020-03-03-SPM-with-parameter-partition.md]: `execute_time` is for a specific value of parameter, the binding may have larger time for this parameter value, but have smaller time for another parameter value?

**Pull request closed without merging**